src/Common/src/System/Net/Sockets/SocketAPMExtensions.cs

```diff
@@ -829,6 +829,24 @@ public static int EndSendTo(this Socket socket, IAsyncResult asyncResult)
         {
             return TaskToApm.End<int>(asyncResult);
         }
+
+        // Behavior adapter.
+        private static ArraySegment<byte> CreateArraySegment(byte[] buffer, int offset, int size)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+
+            try
+            {
+                return new ArraySegment<byte>(buffer, offset, size);
+            }
+            catch (ArgumentException ae)
+            {
+                throw new ArgumentOutOfRangeException(ae.Message, ae);
```

Comment:
You're losing the ParamName value in this translation.  It's a little unfortunate that you have to decide which is worse:
* Losing ParamName
* Discarding InnerException (which might have had more InnerException-ness)
* Rolling back the code to maintain exception parity
* Losing the ArgumentOutOfRangeException